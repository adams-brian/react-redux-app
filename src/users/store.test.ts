import { actionCreators, CREATE_USER, DELETE_USER,
  listReducer, UPDATE_USER, USERS_UPDATED } from "./store";

const user = {
  _id: '123',
  firstname: 'abc',
  lastname: 'def'
}

const list = [
  {
    _id: '456',
    firstname: 'hij',
    lastname: 'klm'
  },
  {
    _id: '789',
    firstname: 'nop',
    lastname: 'qrs'
  }
];

describe('store', () => {
  
  describe('actions', () => {

    it('should export action types', () => {
      expect(CREATE_USER).toMatchSnapshot();
      expect(UPDATE_USER).toMatchSnapshot();
      expect(DELETE_USER).toMatchSnapshot();
      expect(USERS_UPDATED).toMatchSnapshot();
    });

  });

  describe('actionCreators', () => {
    
    it('should export actionCreators', () => {
      expect(actionCreators.createUser(user)).toMatchSnapshot();
      expect(actionCreators.deleteUser(user._id)).toMatchSnapshot();
      expect(actionCreators.updateUser(user)).toMatchSnapshot();
      expect(actionCreators.usersUpdated(list)).toMatchSnapshot();
    });

  });

  describe('listReducer', () => {

    it('should append when creating a user', () => {
      expect(listReducer(undefined, actionCreators.createUser(user))).toEqual([user]);
      expect(listReducer(list, actionCreators.createUser(user))).toEqual([...list, user]);
    });

    it('should replace when updating a user', () => {
      const update = {
        _id: '456',
        firstname: 'updated',
        lastname: 'user'
      };
      expect(listReducer(undefined, actionCreators.updateUser(update))).toEqual([]);
      expect(listReducer(list, actionCreators.updateUser(update))).toEqual(
        list.map(u => u._id === update._id ? Object.assign({}, u, update) : u)
      );
      expect(listReducer(list, actionCreators.updateUser(user))).toEqual(list);
    });

    it('should remove when deleting a user', () => {
      expect(listReducer(undefined, actionCreators.deleteUser('456'))).toEqual([]);
      expect(listReducer(list, actionCreators.deleteUser('456'))).toEqual(
        list.filter(u => u._id !== '456')
      );
      expect(listReducer(list, actionCreators.deleteUser('123'))).toEqual(list);
    });

    it('should replace all when updating users', () => {
      const newList = [
        {
          _id: '654',
          firstname: 'jih',
          lastname: 'mlk'
        },
        {
          _id: '987',
          firstname: 'pon',
          lastname: 'srq'
        }
      ];
      expect(listReducer(undefined, actionCreators.usersUpdated(newList))).toEqual(newList);
      expect(listReducer(list, actionCreators.usersUpdated(newList))).toEqual(newList);
      expect(listReducer(list, actionCreators.usersUpdated([]))).toEqual([]);
    });

  });

});
