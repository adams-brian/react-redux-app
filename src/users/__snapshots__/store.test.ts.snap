// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store actionCreators should export actionCreators 1`] = `
Object {
  "payload": Object {
    "_id": "123",
    "firstname": "abc",
    "lastname": "def",
  },
  "type": "[Users] CREATE",
}
`;

exports[`store actionCreators should export actionCreators 2`] = `
Object {
  "payload": "123",
  "type": "[Users] DELETE",
}
`;

exports[`store actionCreators should export actionCreators 3`] = `
Object {
  "payload": Object {
    "_id": "123",
    "firstname": "abc",
    "lastname": "def",
  },
  "type": "[Users] UPDATE",
}
`;

exports[`store actionCreators should export actionCreators 4`] = `
Object {
  "payload": Array [
    Object {
      "_id": "456",
      "firstname": "hij",
      "lastname": "klm",
    },
    Object {
      "_id": "789",
      "firstname": "nop",
      "lastname": "qrs",
    },
  ],
  "type": "[Users] USERS UPDATED",
}
`;

exports[`store actions should export action types 1`] = `"[Users] CREATE"`;

exports[`store actions should export action types 2`] = `"[Users] UPDATE"`;

exports[`store actions should export action types 3`] = `"[Users] DELETE"`;

exports[`store actions should export action types 4`] = `"[Users] USERS UPDATED"`;
